Review

juancruzgs opened the pull request:
The code looks really good!
I just added some commits with minimal changes, let me know what you think.
There are more places in the code where these improvements can be replicated, feel free to change them in this PR if you are ok with my suggestion.

juancruzgs committed:
Support library for vector drawables not needed

juancruzgs committed:
Use text appearance from material design

juancruzgs committed:
Remove hardcoded strings

juancruzgs committed:
Use a constant for base url

juancruzgs committed:
Inject user repository

juancruzgs committed:
Use two-way data binding

juancruzgs commented on app/src/main/res/layout/fragment_submission_detail.xml:
```diff
@@ -31,7 +31,7 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:text='@{"Reported contact count: " + Integer.toString(viewModel.selectedSubmission.submission.contactCount)}'
+            android:text="@{@string/submission_detail_contact_count(viewModel.selectedSubmission.submission.contactCount)}"
```
All the strings are hardcoded in the xml.
It may be better to move all of them to `strings.xml` to format them and edit them as needed

schmecs replied:
Thanks 👍 Will address updates to the submission detail page including 2-way databinding in a separate branch

juancruzgs commented on app/src/main/res/layout/submission_item.xml:
```diff
@@ -10,7 +10,8 @@
         android:layout_height="wrap_content"
         android:padding="10dp"
-        android:textSize="24sp"
+        android:textAppearance="?attr/textAppearanceHeadline5"
+        android:textColor="?android:textColorSecondary"
```
I recommend you to use material design attributes.
You can change the default values by overriding the style in `styles.xml`.
In that way you won't have harcoded values in the layouts and it will be easier to change in the future

schmecs replied:
Added as an issue to address separately 👍

juancruzgs commented on app/src/main/java/com/example/qstreak/ui/OnboardingSignupFragment.kt:
```diff
@@ -29,24 +29,17 @@ class OnboardingSignupFragment : Fragment() {
-        // TODO data validation - zip required, age and household size optional
-        onboardingViewModel.createUser(
-            requireContext(),
-            binding.ageInput.editText?.text.toString().toInt(),
-            binding.householdSizeInput.editText?.text.toString().toInt(),
-            binding.zipCodeInput.editText?.text.toString()
```
I moved this to the `ViewModel`, it can use two-way databinding to read/write the values

juancruzgs commented on app/src/main/java/com/example/qstreak/ui/OnboardingSignupFragment.kt:
```diff
@@ -29,24 +29,17 @@ class OnboardingSignupFragment : Fragment() {
-            createUser()
+        binding.lifecycleOwner = activity
+        binding.viewModel = onboardingViewModel
+        binding.signUpCallback = {
```
Instead of manually setting the click listener I prefer to send the method reference to the layout

schmecs replied:
Makes sense, that would be more consistent

juancruzgs commented on app/src/main/java/com/example/qstreak/viewmodels/OnboardingViewModel.kt:
```diff
@@ -1,24 +1,33 @@
 
-class OnboardingViewModel(application: Application) : AndroidViewModel(application) {
-    private val userRepository = UserRepository(QstreakDatabase.getInstance(application).userDao())
+class OnboardingViewModel(private val userRepository: UserRepository) : ViewModel() {
```
Inject UserRepository instead of creating it here

juancruzgs commented on app/src/main/java/com/example/qstreak/viewmodels/OnboardingViewModel.kt:
```diff
@@ -1,24 +1,33 @@
     val signupSuccessful = MutableLiveData<Boolean>(false)
 
-    fun createUser(context: Context, age: Int, householdSize: Int, zip: String) {
+    val age = MutableLiveData<String>()
+    val householdSize = MutableLiveData<String>()
+    val zipCode = MutableLiveData<String>()
```
Two-way databinding
https://developer.android.com/topic/libraries/data-binding/two-way

juancruzgs commented on app/src/main/res/layout/fragment_onboarding_signup.xml:
```diff
@@ -33,6 +43,7 @@
                 <com.google.android.material.textfield.TextInputEditText
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
+                    android:text="@={viewModel.zipCode}"
```
Two-way databinding
https://developer.android.com/topic/libraries/data-binding/two-way

schmecs commented on app/src/main/java/com/example/qstreak/viewmodels/OnboardingViewModel.kt:
```diff
@@ -1,24 +1,33 @@
+        val householdSize = householdSize.value
+        val zipCode = zipCode.value
+        if (age.isNullOrBlank() || householdSize.isNullOrBlank() || zipCode.isNullOrBlank()) {
+            // Post a LiveData containing errors to be handled by the Fragment
```
👍

schmecs replied:
Data validation to be addressed in a separate PR

schmecs commented on app/src/main/res/layout/fragment_onboarding_signup.xml:
```diff
@@ -74,6 +87,7 @@
                 android:layout_height="wrap_content"
                 android:layout_margin="60dp"
                 android:text="Continue"
+                android:onClick="@{() -> signUpCallback.invoke()}"
```
If I also take Shalom's advice and refactor so there's no need for a `context`, we could even call `viewModel.createUser` directly here 👍

juancruzgs replied:
💯

schmecs committed:
Remove need for context in Viewmodels
- Injects SharedPreferences and Editor instead of EncryptedSharedPreferences to handle UID logic and remove need for context in ViewModels
- Calls OnboardingViewModel submit function directly via databinding
- Adds SplashViewModel to handle check for UID at app launch
- Extracts onboarding string resources
